File: crop_las_by_polygons/src/bench.zig
const std = @import("std");
const ptinpoly = @import("ptinpoly.zig");
const Point = ptinpoly.Point;
const PreparedPolygon = ptinpoly.PreparedPolygon;

/// Crossings-multiply reference implementation matching the C baseline.
fn naiveContains(vertices: []const Point, point: Point) bool {
    const tx = point.x;
    const ty = point.y;

    var inside = false;
    var previous = vertices[vertices.len - 1];
    var yflag0 = previous.y >= ty;

    for (vertices) |current| {
        const yflag1 = current.y >= ty;
        if (yflag0 != yflag1) {
            const lhs = (current.y - ty) * (previous.x - current.x);
            const rhs = (current.x - tx) * (previous.y - current.y);
            if ((lhs >= rhs) == yflag1) inside = !inside;
        }

        yflag0 = yflag1;
        previous = current;
    }

    return inside;
}

/// Star polygon with `n_tips` arms, alternating between outer_r and inner_r.
fn buildStar(allocator: std.mem.Allocator, n_tips: usize, outer_r: f64, inner_r: f64) ![]Point {
    const n = n_tips * 2;
    const pts = try allocator.alloc(Point, n);
    for (0..n) |i| {
        const angle = std.math.pi * 2.0 * @as(f64, @floatFromInt(i)) / @as(f64, @floatFromInt(n)) - std.math.pi / 2.0;
        const r = if (i % 2 == 0) outer_r else inner_r;
        pts[i] = .{ .x = r * @cos(angle), .y = r * @sin(angle) };
    }
    return pts;
}

const BenchCase = struct {
    name: []const u8,
    polygon: []const Point,
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const stdout = std.fs.File.stdout().deprecatedWriter();

    const N: usize = 1_000_000;

    // Generate fixed random query points in [-1.2, 1.2]^2.
    var prng = std.Random.DefaultPrng.init(0xdeadbeef);
    const rng = prng.random();

    const xs = try allocator.alloc(f64, N);
    defer allocator.free(xs);
    const ys = try allocator.alloc(f64, N);
    defer allocator.free(ys);

    for (xs, ys) |*x, *y| {
        x.* = rng.float(f64) * 2.4 - 1.2;
        y.* = rng.float(f64) * 2.4 - 1.2;
    }

    const square = [_]Point{
        .{ .x = -1.0, .y = -1.0 },
        .{ .x = 1.0, .y = -1.0 },
        .{ .x = 1.0, .y = 1.0 },
        .{ .x = -1.0, .y = 1.0 },
    };

    const star10 = try buildStar(allocator, 10, 1.0, 0.45);
    defer allocator.free(star10);
    const star100 = try buildStar(allocator, 100, 1.0, 0.45);
    defer allocator.free(star100);
    const star500 = try buildStar(allocator, 500, 1.0, 0.45);
    defer allocator.free(star500);

    const cases = [_]BenchCase{
        .{ .name = "square (4 edges)", .polygon = &square },
        .{ .name = "star (20 edges)", .polygon = star10 },
        .{ .name = "star (200 edges)", .polygon = star100 },
        .{ .name = "star (1000 edges)", .polygon = star500 },
    };

    try stdout.print("Benchmarking {d} queries per case\n", .{N});
    try stdout.print("{s:<24} {s:<18} {s:>14}  {s:>10}  {s}\n", .{
        "polygon", "strategy", "ns/query", "prep (us)", "inside",
    });
    try stdout.print("{s}\n", .{"-" ** 86});

    for (cases) |c| {
        // ---- naive ----
        {
            var timer = try std.time.Timer.start();
            var hits: u64 = 0;
            for (0..N) |i| {
                if (naiveContains(c.polygon, .{ .x = xs[i], .y = ys[i] })) hits += 1;
            }
            const ns = timer.read();
            try stdout.print("{s:<24} {s:<18} {d:>14.1}  {s:>10}  {d}\n", .{
                c.name, "naive", nsPerQuery(ns, N), "-", hits,
            });
        }

        // ---- grid, several resolutions ----
        inline for (.{ 4, 8, 16, 32, 64, 128 }) |res| {
            var prep_timer = try std.time.Timer.start();
            var prepared = try PreparedPolygon.init(allocator, c.polygon, res);
            const prep_ns = prep_timer.read();
            defer prepared.deinit();

            // warm-up pass (avoids cold-cache bias on first timing)
            var warmup: u64 = 0;
            for (0..N / 10) |i| {
                if (prepared.contains(.{ .x = xs[i], .y = ys[i] })) warmup += 1;
            }
            if (warmup > N) unreachable; // prevent optimization

            var timer = try std.time.Timer.start();
            var hits: u64 = 0;
            for (0..N) |i| {
                if (prepared.contains(.{ .x = xs[i], .y = ys[i] })) hits += 1;
            }
            const ns = timer.read();
            const label = std.fmt.comptimePrint("grid res={d}", .{res});
            try stdout.print("{s:<24} {s:<18} {d:>14.1}  {d:>10.1}  {d}\n", .{
                "", label, nsPerQuery(ns, N), @as(f64, @floatFromInt(prep_ns)) / 1000.0, hits,
            });
        }

        try stdout.print("\n", .{});
    }
}

fn nsPerQuery(total_ns: u64, n: usize) f64 {
    return @as(f64, @floatFromInt(total_ns)) / @as(f64, @floatFromInt(n));
}

File: crop_las_by_polygons/c_ref/bench.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>
#include "ptinpoly.h"

#define N 1000000

static double xs[N], ys[N];

/* Simple xoshiro128+ PRNG matching Zig's DefaultPrng(0xdeadbeef) output isn't
   practical, so we just use our own deterministic sequence.  The exact inside
   counts will differ from the Zig benchmark, but timings are comparable. */
static unsigned long long rng_state = 0xdeadbeef;
static double rng_f64(void) {
    rng_state ^= rng_state << 13;
    rng_state ^= rng_state >> 7;
    rng_state ^= rng_state << 17;
    return (double)(rng_state & 0xFFFFFFFFFFFFULL) / (double)0xFFFFFFFFFFFFULL;
}

static void gen_star(double (*out)[2], int n_tips, double outer_r, double inner_r) {
    int total = n_tips * 2;
    for (int i = 0; i < total; i++) {
        double angle = M_PI * 2.0 * i / total - M_PI / 2.0;
        double r = (i % 2 == 0) ? outer_r : inner_r;
        out[i][0] = r * cos(angle);
        out[i][1] = r * sin(angle);
    }
}

typedef struct {
    const char *name;
    double (*pgon)[2];
    int nverts;
} BenchCase;

static void bench_naive(BenchCase *c) {
    struct timespec t0, t1;
    unsigned long hits = 0;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < N; i++) {
        double pt[2] = { xs[i], ys[i] };
        if (CrossingsMultiplyTest(c->pgon, c->nverts, pt)) hits++;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double ns = ((t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec)) / N;
    printf("%-24s %-18s %14.1f  %10s  %lu\n", c->name, "C naive", ns, "-", hits);
}

static void bench_grid(BenchCase *c, int res) {
    GridSet gs;
    struct timespec t0, t1;

    clock_gettime(CLOCK_MONOTONIC, &t0);
    GridSetup(c->pgon, c->nverts, res, &gs);
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double prep_us = ((t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec)) / 1000.0;

    /* warm-up */
    volatile unsigned long warmup = 0;
    for (int i = 0; i < N / 10; i++) {
        double pt[2] = { xs[i], ys[i] };
        if (GridTest(&gs, pt)) warmup++;
    }

    unsigned long hits = 0;
    clock_gettime(CLOCK_MONOTONIC, &t0);
    for (int i = 0; i < N; i++) {
        double pt[2] = { xs[i], ys[i] };
        if (GridTest(&gs, pt)) hits++;
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double ns = ((t1.tv_sec - t0.tv_sec) * 1e9 + (t1.tv_nsec - t0.tv_nsec)) / N;

    char label[32];
    snprintf(label, sizeof(label), "C grid res=%d", res);
    printf("%-24s %-18s %14.1f  %10.1f  %lu\n", "", label, ns, prep_us, hits);

    GridCleanup(&gs);
}

int main(void) {
    /* Generate test points in [-1.2, 1.2]^2 */
    for (int i = 0; i < N; i++) {
        xs[i] = rng_f64() * 2.4 - 1.2;
        ys[i] = rng_f64() * 2.4 - 1.2;
    }

    double square[4][2] = {
        {-1.0, -1.0}, {1.0, -1.0}, {1.0, 1.0}, {-1.0, 1.0}
    };

    double star10[20][2];   gen_star(star10, 10, 1.0, 0.45);
    double star100[200][2]; gen_star(star100, 100, 1.0, 0.45);
    double star500[1000][2]; gen_star(star500, 500, 1.0, 0.45);

    BenchCase cases[] = {
        { "square (4 edges)",   square,  4 },
        { "star (20 edges)",    star10,  20 },
        { "star (200 edges)",   star100, 200 },
        { "star (1000 edges)",  star500, 1000 },
    };
    int n_cases = sizeof(cases) / sizeof(cases[0]);

    printf("Benchmarking %d queries per case (original C code)\n", N);
    printf("%-24s %-18s %14s  %10s  %s\n",
        "polygon", "strategy", "ns/query", "prep (us)", "inside");
    for (int i = 0; i < 86; i++) putchar('-');
    putchar('\n');

    int resolutions[] = { 4, 8, 16, 32, 64, 128 };
    int n_res = sizeof(resolutions) / sizeof(resolutions[0]);

    for (int c = 0; c < n_cases; c++) {
        bench_naive(&cases[c]);
        for (int r = 0; r < n_res; r++) {
            bench_grid(&cases[c], resolutions[r]);
        }
        printf("\n");
    }

    return 0;
}
